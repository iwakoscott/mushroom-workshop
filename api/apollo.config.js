module.exports = {
  service: {
    name: 'mushroom-7932'
  }
};
