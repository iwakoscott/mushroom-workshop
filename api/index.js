const { ApolloServer, gql } = require("apollo-server");
const mushroomAPI = require("./mushroomAPI");

// Construct a schema, using GraphQL schema language
const typeDefs = gql`
  type Location {
    name: String!
  }

  type Consensus {
    name: String!
  }

  type Mushroom {
    id: ID!
    date: String!
    latitude: String!
    longitude: String!
    consensus: Consensus
    location: Location
  }

  type Query {
    getMushroomsFrom(
      north: Float
      east: Float
      south: Float
      west: Float
    ): [Mushroom!]!
  }
`;

// Provide resolver functions for your schema fields
const resolvers = {
  Query: {
    getMushroomsFrom: async (
      _,
      { north, east, south, west },
      { dataSources }
    ) => dataSources.mushroomAPI.getMushroomsFrom({ north, east, south, west })
  }
};

const server = new ApolloServer({
  typeDefs,
  dataSources: () => ({
    mushroomAPI: new mushroomAPI()
  }),
  resolvers
});

server.listen().then(({ url }) => {
  console.log(`ğŸš€ Server ready at ${url}`);
});
