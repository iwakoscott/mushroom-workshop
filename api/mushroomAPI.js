const { RESTDataSource } = require("apollo-datasource-rest");

class MushroomAPI extends RESTDataSource {
  constructor() {
    super();
    this.baseURL = "http://mushroomobserver.org/api/";
  }

  mushroomReducer(data) {
    return {
      id: data.id,
      date: data.date,
      latitude: data.latitude,
      longitude: data.longitude,
      consensus: {
        name: data.consensus.name
      },
      location: {
        name: data.location.name
      }
    };
  }

  async getMushroomsFrom({
    north = 38.0461,
    east = -122.801,
    south = 38.0393,
    west = -122.83
  }) {
    const response = await this.get(`observations`, {
      north,
      east,
      south,
      west,
      format: "json",
      detail: "low"
    });

    return response.results.map(this.mushroomReducer);
  }
}

module.exports = MushroomAPI;
