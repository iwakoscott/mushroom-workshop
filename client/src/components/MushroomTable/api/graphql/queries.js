import gql from 'graphql-tag';

export const GET_MUSHROOMS = gql`
  query getMushrooms {
    getMushroomsFrom {
      id
      date
      longitude
      latitude
      consensus {
        name
      }
      location {
        name
      }
    }
  }
`;
