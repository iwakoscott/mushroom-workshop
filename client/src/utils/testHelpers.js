import React from 'react';
import { MockedProvider } from 'react-apollo';
import { render } from 'react-testing-library';

export const renderWithApollo = (
  ui,
  { mocks = [], addTypename = false } = {}
) => ({
  ...render(
    <MockedProvider mocks={mocks} addTypename={addTypename}>
      {ui}
    </MockedProvider>
  )
});

export const providerOptions = ({
  query,
  variables,
  data,
  addTypename = false
}) => ({
  mocks: [
    {
      request: {
        query,
        variables
      },
      result: {
        data
      }
    }
  ],
  addTypename
});
