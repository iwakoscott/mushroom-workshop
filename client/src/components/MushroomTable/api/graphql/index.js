import { GET_MUSHROOMS } from "./queries";

export { GET_MUSHROOMS };