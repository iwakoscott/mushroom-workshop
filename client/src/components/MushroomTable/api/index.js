import { graphql } from 'react-apollo';
import { GET_MUSHROOMS } from './graphql';

export const withMushrooms = graphql(GET_MUSHROOMS);
