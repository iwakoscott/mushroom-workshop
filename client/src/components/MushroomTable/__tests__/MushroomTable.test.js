import React from 'react';
import { fireEvent, wait } from 'react-testing-library';
import { renderWithApollo, providerOptions } from '../../../utils/testHelpers';
import { MushroomTable } from '../';
import { GET_MUSHROOMS } from '../api/graphql';

describe(`<MushroomTable />`, () => {
  const fakeDataA = {
    id: '81732',
    date: '2011-11-06',
    longitude: '-122.8045',
    latitude: '38.0452',
    consensus: {
      name: 'Phaeoclavulina curta'
    },
    location: {
      name: 'Point Reyes National Seashore, Marin Co., California, USA'
    }
  };

  const fakeDataB = {
    id: '265941',
    date: '2016-12-31',
    longitude: '-122.805',
    latitude: '38.0456',
    consensus: {
      name: 'Entoloma subgenus Cyanula'
    },
    location: {
      name:
        'Horse Trail, Point Reyes National Seashore, Point Reyes Station, California, USA'
    }
  };

  const getMushroomsFrom = [fakeDataA, fakeDataB];

  const getProviderOptions = () =>
    providerOptions({
      query: GET_MUSHROOMS,
      data: {
        getMushroomsFrom
      }
    });

  it(`renders without crashing...`, async () => {
    const options = getProviderOptions();
    const { container } = renderWithApollo(<MushroomTable />, options);
    await wait(() => 'flush promise');
    expect(container).toMatchSnapshot();
  });

  it(`displays the data in a table`, async () => {
    const options = getProviderOptions();
    const { container } = renderWithApollo(<MushroomTable />, options);

    await wait(() => 'flush promises');

    // DOM Nodes
    const tableRowNodes = container.querySelectorAll('tr');

    expect(tableRowNodes).toHaveLength(getMushroomsFrom.length + 1); // includes the header

    // since we are turning the data into a table we expect all data properties to be in the table
    getMushroomsFrom.forEach(data => {
      Object.values(data).forEach(value =>
        expect(container.innerHTML).toMatch(new RegExp(value))
      );
    });
  });
});
