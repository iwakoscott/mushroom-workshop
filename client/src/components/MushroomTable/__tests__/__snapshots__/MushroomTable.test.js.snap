// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MushroomTable /> renders without crashing... 1`] = `
<div>
  <div
    class="MuiPaper-root-12 MuiPaper-elevation1-15 MuiPaper-rounded-13 MuiCard-root-11"
    style="padding: 1em; margin: 1em;"
  >
    <table
      class="MuiTable-root-39"
    >
      <thead
        class="MuiTableHead-root-40"
      >
        <tr
          class="MuiTableRow-root-41 MuiTableRow-head-44"
        >
          <th
            class="MuiTableCell-root-46 MuiTableCell-head-47"
            scope="col"
          >
            id
          </th>
          <th
            class="MuiTableCell-root-46 MuiTableCell-head-47"
            scope="col"
          >
            Mushroom Name
          </th>
          <th
            class="MuiTableCell-root-46 MuiTableCell-head-47"
            scope="col"
          >
            Date Posted
          </th>
          <th
            class="MuiTableCell-root-46 MuiTableCell-head-47"
            scope="col"
          >
            Longitude
          </th>
          <th
            class="MuiTableCell-root-46 MuiTableCell-head-47"
            scope="col"
          >
            Latitude
          </th>
          <th
            class="MuiTableCell-root-46 MuiTableCell-head-47"
            scope="col"
          >
            Location
          </th>
        </tr>
      </thead>
      <tbody
        class="MuiTableBody-root-58"
      >
        <tr
          class="MuiTableRow-root-41 MuiTableRow-hover-43"
        >
          <td
            class="MuiTableCell-root-46 MuiTableCell-body-48"
          >
            81732
          </td>
          <td
            class="MuiTableCell-root-46 MuiTableCell-body-48"
          >
            Phaeoclavulina curta
          </td>
          <td
            class="MuiTableCell-root-46 MuiTableCell-body-48"
          >
            2011-11-06
          </td>
          <td
            class="MuiTableCell-root-46 MuiTableCell-body-48"
          >
            -122.8045
          </td>
          <td
            class="MuiTableCell-root-46 MuiTableCell-body-48"
          >
            38.0452
          </td>
          <td
            class="MuiTableCell-root-46 MuiTableCell-body-48"
          >
            Point Reyes National Seashore, Marin Co., California, USA
          </td>
        </tr>
        <tr
          class="MuiTableRow-root-41 MuiTableRow-hover-43"
        >
          <td
            class="MuiTableCell-root-46 MuiTableCell-body-48"
          >
            265941
          </td>
          <td
            class="MuiTableCell-root-46 MuiTableCell-body-48"
          >
            Entoloma subgenus Cyanula
          </td>
          <td
            class="MuiTableCell-root-46 MuiTableCell-body-48"
          >
            2016-12-31
          </td>
          <td
            class="MuiTableCell-root-46 MuiTableCell-body-48"
          >
            -122.805
          </td>
          <td
            class="MuiTableCell-root-46 MuiTableCell-body-48"
          >
            38.0456
          </td>
          <td
            class="MuiTableCell-root-46 MuiTableCell-body-48"
          >
            Horse Trail, Point Reyes National Seashore, Point Reyes Station, California, USA
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
